shared/js/advanced_telemetry_helper.js

```diff
@@ -0,0 +1,203 @@
+ *  count.add(); //Increment the count. The count of the histogram is now 1.
+ *  // DO SOMETHING
+ *  count.add(); //Increment the count. The count of the histogram is now 2.
+ *
+ * Linear Metric Usage Example:
+ *  // This creates a linear histogram called 'mylinear' with minimum
+ *  // of 1 (0 not allowed as a minimum), max of 1000, with 10 buckets.
+ *  var linear = new AdvancedTelemetryHelper(ATH.HISTOGRAM_LINEAR, 'mylinear',
+ *      1, 1000, 10);
+ *  // This adds the value 15 to the histogram 'mylinear'
+ *  linear.add('mylinear', 15);
+ *  // DO SOMETHING
+ *  // This adds to the existing histogram a value of 800.
+ *  linear.add(800); // The histogram now has two values, one in the 15 bucket
+ *                   // and one in the bucket that holds the value 800
+ *
+ * Exponential Metric Usage Example:
+ *  // This creates an exponential histogram called 'myexp' with minimum
+ *  // of 1 (0 not allowed as a minimum), max of 1000, with 10 buckets.
+ *  var exp = new AdvancedTelemetryHelper(ATH.HISTOGRAM_EXPONENTIAL, 'myexp',
+ *      1, 1000, 10);
+ *  // This adds the value 15 to the histogram 'myexp'
+ *  exp.add('myexp', 15);
+ *  // DO SOMETHING
+ *  // This adds to the existing histogram a value of 800.
+ *  exp.add(800); // The histogram now has two values, one in the 15 bucket
+ *                // and one in the bucket that holds the value 800
+ *
+ * Caveats:
+ *  Note that after a histogram is created, it's not possible to change the type
+ *  of the histogram from one type to the other.  It's necessary to create a new
+ *  Histogram.
+ *
+ *  Each histogram should have a unique name.
+ *
+ */
+
+/* globals SettingsListener */
+(function(exports) {
+  'use strict';
+
+  const APP_TELEMETRY_LOG_PREFIX = 'telemetry';
+  const ATH = AdvancedTelemetryHelper;
+  // Export Histogram types so they can be used by apps
+  ATH.HISTOGRAM_COUNT = 4;
+  ATH.HISTOGRAM_EXP = 0;
+  ATH.HISTOGRAM_LINEAR = 1;
+  // This is exposed for unit testing purposes
+  ATH.TELEMETRY_ENABLED_KEY = 'debug.performance_data.advanced_telemetry';
+  ATH.SETTINGS_INITIALIZED = false;
+
+  // API for an exponential histogram entry.  This function creates the
+  // histogram in the system and prepares it subsequent .log(value) calls
+  // against it.
+  // <type>: one of ATH.HISTOGRAM_COUNT, ATH.HISTOGRAM_EXP,
+  //        or ATH.HISTOGRAM_LINEAR.
+  // <name>: Unique name of the metric
+  // <min>: the minimum (non-zero) numeric value for this histogram.
+  // <max>: the maximum numeric value for this histogram
+  // <buckets>: the total number of buckets for this histogram.
+  function AdvancedTelemetryHelper(type, name, min, max, buckets) {
+    if (!ATH.SETTINGS_INITIALIZED) {
+      ATH.SETTINGS_INITIALIZED = true;
+      ATH.init();
+    }
+
+    if (typeof type !== 'undefined' && typeof name !== 'undefined') {
+      if (type == ATH.HISTOGRAM_COUNT) {
+        this.count(name);
```

Comment:
It seems possible to remove the 'count' function and use createHistogram here, passing '0' for min, max, and buckets.